app/Http/Controllers/Api/ImageAnnotationController.php

```diff
@@ -331,4 +357,108 @@ public function destroy($id)
 
         return response('Deleted.', 200);
     }
+
+    /**
+     * Get all label trees that are used by all projects which are visible to the user.
+     *
+     * @param mixed $user
+     * @param int $volumeId
+     *
+     * @return array
+     */
+    protected function getLabelTreeIds($user, $volumeId)
+    {
+        if ($user->can('sudo')) {
+            // Global admins have no restrictions.
+            $projectIds = DB::table('project_volume')
+                ->where('volume_id', $volumeId)
+                ->pluck('project_id');
+        } else {
+            // Array of all project IDs that the user and the image have in common
+            // and where the user is editor, expert or admin.
+            $projectIds = Project::inCommon($user, $volumeId, [
+                Role::editorId(),
+                Role::expertId(),
+                Role::adminId(),
+            ])->pluck('id');
+        }
+        $trees = LabelTree::select('id', 'name', 'version_id')
+            ->with('labels', 'version')
+            ->whereIn('id', function ($query) use ($projectIds) {
+                $query->select('label_tree_id')
+                    ->from('label_tree_project')
+                    ->whereIn('project_id', $projectIds);
+            })
+            ->pluck('id')
+            ->toArray();
+
+        return $trees;
+    }
+
+    /**
+     * Perform ANN search (HNSW + Post-Subquery-Filtering).
+     *
+     * @param vector $featureVector
+     * @param int[] $trees
+     *
+     * @return array
+     */
+    protected function performAnnSearch($featureVector, $trees)
+    {
+        $featureVector = new Vector($featureVector);
+
+        $subquery = ImageAnnotationLabelFeatureVector::select('label_id', 'label_tree_id')
+            ->selectRaw('(vector <=> ?) AS distance', [$featureVector])
+            ->orderBy('distance')
+            // K = 100
+            ->limit(config('labelbot.K'));
+
+        return DB::table(DB::raw("({$subquery->toSql()}) as subquery"))
+            ->setBindings([$featureVector])
+            ->whereIn('label_tree_id', $trees)
+            ->select('label_id')
+            ->groupBy('label_id')
+            ->orderByRaw('MIN(distance)')
+            ->limit(config('labelbot.N')) // N = 3
+            ->pluck('label_id')
+            ->toArray();
```

Comment:
Does this produce the same SQL query than you developed in your thesis? You can check by placing this at the beginning of the method:
```php 
DB::listen(fn ($q) => dd($q->sql));
```
This will dump the SQL when you run a test.

Reply:
Yes it does. 

ANN:

"select "label_id" 
from (
    select "label_id", "label_tree_id", (vector <=> ?) AS distance 
    from "image_annotation_label_feature_vectors" 
    order by "distance" asc 
    limit 100
    ) as "subquery" 
where "label_tree_id" in (?) 
group by "label_id" 
order by MIN(distance) 
limit 3"


KNN:

"select "label_id" 
from (
    select "label_id", "label_tree_id", (vector <=> ?) AS distance 
    from "image_annotation_label_feature_vectors" 
    where "label_tree_id" in (?) 
    order by "distance" asc 
    limit 100
    ) as "subquery" 
group by "label_id" 
order by MIN(distance) 
limit 3"